engine/src/main/scala/cromwell/engine/workflow/profiler/WorkflowProfiler.scala

```diff
@@ -0,0 +1,95 @@
+  }
+}
+
+case class WorkflowProfiler(workflowId: WorkflowId, serviceRegistryActor: ActorRef) extends Actor with ActorLogging {
+
+  import WorkflowProfiler._
+
+  private val tag = s"WorkflowProfiler-${workflowId.toString}"
+  implicit val actorSystem = context.system
+  implicit val ec = context.dispatcher
+
+  override def preStart(): Unit = {
+    context.system.eventStream.subscribe(self, classOf[CromwellFsmStateAndData])
+  }
+
+  override def receive = LoggingReceive {
+    case workflowManagerActorMessage: WorkflowManagerMetadataMessage => handleMessageFromWorkflowManagerActor(workflowManagerActorMessage)
+    case workflowActorMessage: WorkflowActorStateAndData => handleMessageFromWorkflowActor(workflowActorMessage)
+    case workflowExecutionActorMessage: WorkflowExecutionActorStateAndData => handleMessageFromWorkflowExecutionActor(workflowExecutionActorMessage)
+    case unhandledMessage: Any => log.warning(s"$tag Unknown Message received: $unhandledMessage")
+  }
+
+  private def handleMessageFromWorkflowManagerActor(workflowManagerMetadataMessage: WorkflowManagerMetadataMessage): Unit = {
+    val metadataEvents = List(
+      MetadataEvent(MetadataKey(workflowId, None, WorkflowMetadataKeys.Id), MetadataValue(workflowManagerMetadataMessage.workflowId)),
+      MetadataEvent(MetadataKey(workflowId, None, WorkflowMetadataKeys.SubmissionTime), MetadataValue(workflowManagerMetadataMessage.submissionTime)),
+      // Currently, submission time is the same as start time
+      MetadataEvent(MetadataKey(workflowId, None, WorkflowMetadataKeys.StartTime), MetadataValue(workflowManagerMetadataMessage.submissionTime))
+    )
+    metadataEvents foreach serviceRegistryActor.publishToMetadataService
+  }
+
+  private def handleMessageFromWorkflowActor(workflowActorStateAndData: WorkflowActorStateAndData): Unit = {
+    workflowActorStateAndData match {
+      case WorkflowActorStateAndData(state, data) if state == InitializingWorkflowState =>
+        // `get` is safe since in this state we are guaranteed to have a ValidWorkflowDescriptor
+        val metadataEventMsg = MetadataEvent(MetadataKey(workflowId, None, WorkflowMetadataKeys.Name), MetadataValue(data.workflowDescriptor.get.name))
+        serviceRegistryActor.publishToMetadataService(metadataEventMsg)
+      case WorkflowActorStateAndData(state, data) =>
+        // This updates the workflow status
+        publishCurrentStateToMetadataService(state.workflowState)
+        if (state.terminal) {
+          // Add the end time of the workflow in the MetadataService
+          val metadataEventMsg = MetadataEvent(MetadataKey(workflowId, None, WorkflowMetadataKeys.EndTime), MetadataValue(DateTime.now.toString))
+          serviceRegistryActor.publishToMetadataService(metadataEventMsg)
+          // As the state is terminal, this is expected to be the last message related to a particular Workflow. This actor's job here seems to be done
+          self ! PoisonPill
```

Comment:
This mechanism to kill itself seems a little fishy to me. It should wait for in-flight updates happening to metadata, isn't it? 😕